README.md
# Agent-based modelling to explore the combined effects of social and demographic changes on rural settlement patterns in North-Western Europe during the Middle Ages (800 CE to 1100 CE).

Around 800 CE, the Carolingian Empire achieved its greatest territorial extent: it covered a large part of Europe, from the Ebro to the Elbe. At a regional scale, settlement patterns mixed small dispersed villages (which corresponded to a loose gathering of few houses) and isolated farms (i. e. agricultural holdings). Villages and farms were not permanent: their location often changed (every one or two centuries) from about tens meters to several hundred meters. At this time period, towns were inherited from the Antiquity. They had a central function as episcopal see but their economic and administrative role was not really important.

Around 1100 CE, regional settlement patterns are composed of enduring population clusters located around the castles and the churches. Population clusters (hamlets, villages, towns) are much more numerous than in 800 CE. Towns that were existing in 800 CE are much larger since they have created productive and commercial activities; new nuclei have emerged around suburban monasteries. Many small towns have emerged; they are most often associated to priories having their own outer wall.

**Thus regional settlements patterns that were dispersed in 800 CE became much more concentrated and hierarchical in 1100 CE.** This phenomenon occured in all regions in North-Western Europe 800 CE to 1100 CE but the intensity of the concentration and the hierarchisation of the settlement pattern can be notably variable across the regions.

**Several processes jointly explain this major transition.**

* The piecemeal dismantling of the Carolingian Empire and the dissipation of powers induced  struggles among lords and thus a rise in violence, even increased by Viking invasions. The result of this was the creation of castles as well as an increase of the need of protection for peasant households;

* The so-called "feudal revolution" of the 11th century during which ever smaller lords appropriated various administrative, fiscal and judicial rights for themselves and their vassals. The need of lords for money was largely created by the militarisation of the society mentioned above. As a result, usage fees and rents paid by peasant households to the lords increased a lot.

* The increase of the religious control of the society from which resulted a great upwelling of piety and level of devotion, an increase of the number of parish churches and the increase of usage fees and rents paid by peasant households to the Church.

* The creation of villager communities by peasant households to increase their productivity and counterbalance the power of lords.

**The agent-based model SimFeodal simulates the emergence of enduring population clusters located around the castles and the churches in a rural region during the Middle Ages.**  In the course of time, population clusters (hamlets, villages, small towns) become more numerous; many of them grow (i. e. their number of peasant households increases) but some clusters grow more than others.

### Modelled entities

| Name  | Category | Type  | Activity | Level |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| Peasant household  | Moving locally and at long distances | Socio-spatial | Active | Micro-geographic |
| Peasant household  |Moving only locally | Socio-spatial | Active | Micro-geographic |
| Lord  | Liege lord | Social | Active | Micro-geographic |
| Lord  | Small lord (enfeoffed vassal, knight...) | Socio-spatial | Active | Micro-geographic |
| Taxing area  | Rents | Spatial | Passive | Micro-geographic |
| Taxing area  | High justice rights | Spatial | Passive | Micro-geographic |
| Taxing area  | Medium and low justice rights | Spatial | Passive | Micro-geographic |
| Taxing area  | Usage rights | Spatial | Passive | Micro-geographic |
| Attractive point  | Parish churches |  Socio-spatial | Active | Micro-geographic |
| Attractive point  | Church without parish rights |  Spatial | Passive | Micro-geographic |
| Attractive point  | Castle |  Spatial | Passive | Micro-geographic |
| Attractive point  | Villager community |  Spatial | Passive | Micro-geographic |
| Population cluster  | Village |  Spatial | Passive | Meso-geographic |
| Population cluster  | Agglomeration (market town, small town) |  Spatial | Passive | Meso-geographic |
| Attraction centre  |  |  Spatial | Passive | Meso-geographic |

### Relocation process of a peasant household within SimFeodal

![Relocation process of a peasant household](PH-movingrules.png)


NEWS.md
### SimFeodal 6.5.1 (2019-07-05)
Commit : [2227c25](https://github.com/SimFeodal/SimFeodal/commit/2227c25)

* Modification de `rayon_migration_locale_fp` pour que ce paramètre accepte une map (évolution de la valeur au cours de la simulation) :
	- param `rayon_migration_locale_fp` passe de `<int>` à `<map<int,int>>`
	- ajout variable globale `rayon_migration_locale_fp_actuel`
	- ajout mécanisme update `rayon_migration_locale_fp_actuel`
	- modification des appels à `rayon_migration_locale_fp` par `rayon_migration_locale_fp_actuel`

* Nouveaux scénarios
	- Exp\_6\_5\_1\_Obj50k : Objectif 50k FP en fin de simulation, 2 variantes :
		- `init_nb_total_fp` = 50 000 et `croissance_demo` = 0 (population stable, arrivée = 50k) 
		- `init_nb_total_fp` = 4 000 et `croissance_demo` =  14.22% (décuplement population, arrivée \~ 50k)
	- Exp\_6\_5\_1\_RayonLocal : Objectifs 40 et 50k en fin de simulation, avec (x10) et sans croissance, variation `rayon_migration_locale_fp`, 4 variantes :
		- `init_nb_total_fp` = 40 000 et `croissance_demo` = 0 (population stable, arrivée = 40k) + `rayon_migration_locale_fp` = [800::2500,900::4000,1000::6000]
		- `init_nb_total_fp` = 4 000 et `croissance_demo` =  12.89% (décuplement population, arrivée ~ 40k) + `rayon_migration_locale_fp` = [800::2500,900::4000,1000::6000]
		- `init_nb_total_fp` = 50 000 et `croissance_demo` = 0% (population stable, arrivée ~50.1k) + `rayon_migration_locale_fp` = [800::2500,900::4000,1000::6000]
		- `init_nb_total_fp` = 4 000 et `croissance_demo` = 14.22% (décuplement population, arrivée ~50.1k) + `rayon_migration_locale_fp` = [800::2500,900::4000,1000::6000]

## SimFeodal 6.5 (2019-07-04)
Commit : [c546d92](https://github.com/SimFeodal/SimFeodal/commit/c546d92)

* Calibration châteaux finie :
	- `nb_tirages_chateaux_gs` = 3
	- `proba_construction_chateau_ps` = 0.5
* Test de nouveaux scénarios :
	- Exp\_6\_5\_Obj40k : Objectif 40k FP en fin de simulation, avec 3 variantes
		- `init_nb_total_fp` = 40 000 et `croissance_demo` = 0 (population stable, arrivée = 40k) 
		- `init_nb_total_fp` = 13 500 et `croissance_demo` = 5.89% (triplement population, arrivée \~40.1k)
 		- `init_nb_total_fp` = 4 000 et `croissance_demo` =  12.89% (décuplement population, arrivée \~ 40)
	- Exp\_6\_5\_Obj50k : Objectif 50k FP en fin de simulation, 1 variante :
		- `init_nb_total_fp` = 17 000 et `croissance_demo` = 5.85% (triplement population, arrivée \~50.1k)

### SimFeodal 6.4.3 (2019-06-14)
Commit : [040c855](https://github.com/SimFeodal/SimFeodal/commit/040c855)

* On teste avec :
	- `nb_tirages_chateaux_gs` = 3 (produit entre 35 et 50 châteaux de GS)
	- `proba_construction_chateau_ps` dans \[0.4, 0.5, 0.6, 0.7, 0.8\] (0.8 produit \~10 châteaux de PS)

### SimFeodal 6.4.2 (2019-06-13)
Commit : [ea2e37e](https://github.com/SimFeodal/SimFeodal/commit/ea2e37e)

* Modification construction château :
	- PS : idem 6.4.1
	- GS : paramètre `nb_tirages_chateaux_gs`, puis `loop times: nb_tirages {if (espace_dispo_chateaux != nil and flip(proba_creer_chateau)) [...construction...]}`
	- Paramètre `proba_construction_chateau_ps` changé à 0.8


### SimFeodal 6.4.1 (2019-05-31)
Commit : [2e3581c](https://github.com/SimFeodal/SimFeodal/commit/2e3581c)

* Modification construction château :
	- GS : `puissance/Sum(puissance_GS) * rnd(min = nb_max_chateaux_par_tour_gs, max = nb_max_chateaux_par_tour_gs+1, step = 1)`
	- PS : `proba_chateau_PS` (\~0.7) + lotterie pondéré (`puissance`) parmi les PS pour savoir qui le construit

## SimFeodal 6.4.0 (2019-05-26)
Commit : 0b3fc3f(https://github.com/SimFeodal/SimFeodal/commit/0b3fc3f)

* Ajout du mécanisme de construction de châteaux `construction_chateau_alternate` : logique de `puissance_i/ Sum(puissance_ i+1)`


## SimFeodal 6.3.0 (2019-05-12)
Commit : [88860b5](https://github.com/SimFeodal/SimFeodal/commit/88860b5)

* Renommage de paramètres et nouvelles valeurs par défaut
* On bloque le nouveau mécanisme de construction de châteaux (\*1.25\*2 et \*7)
* Correction calcul `s_religieuse`
* Ajout outputs pour les seigneurs
* Ajout des scénarios principaux : PopGrowth vs PopInit
* Ajout des scénarios "Objectif50k"

## SimFeodal 6.2.0 (2019-03-07)
Commit : [e377f0b](https://github.com/SimFeodal/SimFeodal/commit/e377f0b)

* Test et calibrage de `nb_requis_paroissiens_insatisfaits`
* Ajout params nb max châteaux par tour
* Test `seuil_creation_paroisse`

## SimFeodal 6.1.0 (2019-01-09) 
Commit : [2891f35](https://github.com/SimFeodal/SimFeodal/commit/2891f35)

* Simplification du modèle (seigneurs, châteaux)
* Optimisation requêtes spatiales (`at_distance` remplacées par `inside (self.shape + distance)`)
* Ajout calcul espace disponible construction chateau (proba * 1.25 / 2 tirages GS + proba * 7 / 1 tirage PS)
* Nouvelle règle création châteaux
* Ajout châteaux dans les outputs
* Renommage de paramètres

# SimFeodal 6.0.0 (2019-01-09)
Commit : [397bcba](https://github.com/SimFeodal/SimFeodal/commit/397bcba)

* Optimisation performances modèle
* Renommage des paramètres
* Paramétrage de valeurs en brut
* Modification de valeurs de paramètres

parameters_valeurs.R
library(tidyverse)
library(glue)
# sensib_params <- tibble(
#   param = character(),
#   type = character(),
#   valeur = character()
# )

distance_detection_agregats <- seq(from = 50, to = 300, length.out = 10) %>%
                                    round() %>%
                                    as.integer()
  

nombre_FP_agregat <- seq(from = 3, to = 25) %>%
  round() %>%
  as.integer()

nombre_agglos_antiques <- seq(from = 0, to = 20) %>%
  round() %>%
  as.integer()

nombre_villages <- seq(from = 5, to = 100, by = 5) %>%
  round() %>%
  as.integer()

nombre_FP_village <- seq(from = 5, to = 25) %>%
  round() %>%
  as.integer()

puissance_communautes <- seq(from = 0, to = 1, by = .05)

apparition_communautes <- seq(from = 800, to = 1100, by = 20) %>%
  round() %>%
  as.integer()

proba_apparition_communaute <- seq(from = 0, to = 1, by = .05)

nombre_foyers_paysans <- seq(from = 1000, to = 10000, by = 500) %>%
  round() %>%
  as.integer()

taux_renouvellement <- seq(from = 0, to = 1, by = .05)

taux_augmentation_FP <- seq(from = 0, to = 0.05, length.out = 20)

taux_mobilite <- seq(from = 0, to = 1, by = .05)

distance_max_dem_local <- seq(from = 500, to = 10000, by = 500) %>%
  round() %>%
  as.integer()

seuil_puissance_armee <- seq(from = 0, to = 1000, by = 50) %>%
  round() %>%
  as.integer()

proba_ponderee_deplacement_lointain <- seq(from = 0, to = 1, by = 0.05)

coef_redevances <- seq(from = 0, to = 40, by = 2) %>%
  round() %>%
  as.integer()

### FIXME : serfs_mobiles <- c(TRUE, FALSE)

# list<int> seuils_distance_max_dem_local <- [2500, 4000, 6000];

nombre_seigneurs_objectif <- seq(from = 50, to = 450, by = 20) %>%
  round() %>%
  as.integer()


nombre_grands_seigneurs <- seq(from = 0, to = 2, by = 1) %>%
  round() %>%
  as.integer()

nombre_petits_seigneurs <- seq(from = 0, to = 50, by = 2) %>%
  round() %>%
  as.integer()

proba_collecter_loyer <- seq(from = 0, to = .5, by = 0.025)

proba_creation_ZP_banaux <- seq(from = 0, to = .5, by = 0.025)

proba_creation_ZP_basseMoyenneJustice <- seq(from = 0, to = .5, by = 0.025)

rayon_min_PS <- seq(from = 100, to = 2000, by = 100) %>%
  round() %>%
  as.integer()

rayon_max_PS <- seq(from = 1000, to = 10000, by = 500) %>%
  round() %>%
  as.integer()

min_fourchette_loyers_PS <- seq(0, .5,  by = .025)
max_fourchette_loyers_PS <- seq(0.25, 1,  by = .05)

proba_don_partie_ZP <- seq(0, 1,  by = .05)

apparition_chateaux <- seq(from = 800, to = 1160, by = 20) %>%
  round() %>%
  as.integer()

nb_chateaux_potentiels_GS <- seq(from = 0, to = 15, by = 1) %>%
  round() %>%
  as.integer()

seuil_attractivite_chateau <- seq(from = 1000, to = 10000, by = 500) %>%
  round() %>%
  as.integer()

proba_creer_chateau_GS <- seq(0, 1,  by = .05)

proba_chateau_agregat <- seq(0, 1,  by = .05)

proba_don_chateau_GS <- seq(0, 1,  by = .05)

proba_creer_chateau_PS <- seq(0, 1,  by = .05)

proba_gain_droits_hauteJustice_chateau <- seq(0, 1,  by = .05)
proba_gain_droits_banaux_chateau <- seq(0, 1,  by = .05)
proba_gain_droits_basseMoyenneJustice_chateau <- seq(0, 1,  by = .05)

proba_promotion_groschateau_multipole <- seq(0, 1,  by = .05)
proba_promotion_groschateau_autre <- seq(0, 1,  by = .05)

puissance_necessaire_creation_chateau_GS <- seq(from = 0, to = 5000, by = 250) %>%
  round() %>%
  as.integer()

puissance_necessaire_creation_chateau_PS <- seq(from = 0, to = 1000, by = 50) %>%
  round() %>%
  as.integer()



nombre_eglises <- seq(from = 0, to = 500, by = 25) %>%
  round() %>%
  as.integer()

nb_eglises_paroissiales <- seq(from = 0, to = 500, by = 25) %>%
  round() %>%
  as.integer()


nb_max_paroissiens <- seq(from = 10, to = 200, by = 10) %>%
  round() %>%
  as.integer()

nb_min_paroissiens <- seq(from = 0, to = 40, by = 5) %>%
  round() %>%
  as.integer()
  
seuil_creation_paroisse <- seq(from = 0, to = 500, by = 25) %>%
  round() %>%
  as.integer()

nb_paroissiens_mecontents_necessaires <- seq(from = 0, to = 100, by = 5) %>%
  round() %>%
  as.integer()

variables <- ls()
variables <- variables[variables != "variables"]

param_name_to_tibble <- function(string){
  name <- string
  type <- class(eval(parse(text = name)))
  values <- eval(parse(text = name))
  tibble(param = name, type = type, valeur = values)
}

types_gama <- tibble(
  type = c("integer","numeric"),
  type_gama = c("INT", "FLOAT")
)

sensib_params <- lapply(X = variables, FUN = param_name_to_tibble) %>%
  bind_rows() %>%
  mutate(id = 1:nrow(.)) %>%
  left_join(types_gama, by = "type") %>%
  mutate(sensibility_value = as.character(valeur)) %>%
  mutate(seed = 1)
  

vars <- ls()
rm(list = vars[vars != "sensib_params"])

xml_header <- '<?xml version="1.0" encoding="UTF-8"?>\n<Experiment_plan>'
xml_content <- '<Simulation id="{id}" sourcePath="/home/robin/transition8/models/Base4_5_GUI.gaml" finalStep="18" experiment="Exp_4_5_sensibility" seed="{seed}">
      <Parameters>
        <Parameter var="{param}" type="{type_gama}" value="{valeur}" />
        <Parameter var="sensibility_parameter" type="STRING" value="{param}" />
        <Parameter var="sensibility_value" type="STRING" value="{sensibility_value}" />
      </Parameters>
      <Outputs>
      </Outputs>
</Simulation>'


for (i in 1:10) {
  sensib_params <- sensib_params %>%
    mutate(seed = i)
  
  simulations_xml <- sensib_params %>%
    glue_data(xml_content) %>%
    paste(collapse = "\n")
  
  final_xml <- paste(xml_header, simulations_xml, sep = "\n")
  final_xml <- paste(final_xml, "</Experiment_plan>", sep = "\n")
  final_xml <- paste(final_xml, "\n")
  writeLines(final_xml, con = glue("simFEODAL_analyse_sensibilite_{i}.xml"))
}

# 
# // POLES //
#   float attrac_0_eglises <- 0.0;
# float attrac_1_eglises <- 0.15;
# float attrac_2_eglises <- 0.25;
# float attrac_3_eglises <- 0.5;
# float attrac_4_eglises <- 0.6;
# float attrac_GC <- 0.25;
# float attrac_PC <- 0.15;
# float attrac_communautes <- 0.15;
